data/helpers.d/psql

```diff
@@ -0,0 +1,88 @@
+
+# Open a connection as a user
+#
+# example: ynh_psql_connect_as 'user' 'pass' <<< "UPDATE ...;"
+# example: ynh_psql_connect_as 'user' 'pass' < /path/to/file.sql
+#
+# usage: ynh_psql_connect_as user pwd [db]
+# | arg: user - the user name to connect as
+# | arg: pwd - the user password
+# | arg: db - the database to connect to
+ynh_psql_connect_as() {
+    ynh_die "ynh_psql_connect_as is not yet implemented"
+}
+
+# # Execute a command as root user
+#
+# usage: ynh_psql_execute_as_root sql [db]
+# | arg: sql - the SQL command to execute
+# | arg: db - the database to connect to
+ynh_psql_execute_as_root () {
+        sudo su -c "psql" - postgres <<< ${1}
+#TODO support db argument ?
+}
+
+# Execute a command from a file as root user
+#
+# usage: ynh_psql_execute_file_as_root file [db]
+# | arg: file - the file containing SQL commands
+# | arg: db - the database to connect to
+ynh_psql_execute_file_as_root() {
+     ynh_die "ynh_psql_execute_file_as_root is not yet implemented"
+}
+
+# Create a database and grant optionnaly privilegies to a user
+#
+# usage: ynh_psql_create_db db [user [pwd]]
+# | arg: db - the database name to create
+# | arg: user - the user to grant privilegies
+# | arg: pwd - the password to identify user by
+ynh_psql_create_db() {
+    db=$1
+    # grant all privilegies to user
+    if [[ $# -gt 1 ]]; then
+        ynh_psql_create_user ${2} "${3}"
+        sudo su -c "createdb -O ${2} $db" -  postgres
+    else
+        sudo su -c "createdb $db" -  postgres
+    fi
+
+}
+
+# Drop a database
+#
+# usage: ynh_psql_drop_db db
+# | arg: db - the database name to drop
+ynh_psql_drop_db() {
+    sudo su -c "dropdb ${1}" -  postgres
+}
+
+# Dump a database
+#
+# example: ynh_psql_dump_db 'roundcube' > ./dump.sql
+#
+# usage: ynh_psql_dump_db db
+# | arg: db - the database name to dump
+# | ret: the psqldump output
+ynh_psql_dump_db() {
+    ynh_die "ynh_psql_dump_db is not yet implemented"
+}
+
+
+# Create a user
+#
+# usage: ynh_psql_create_user user pwd [host]
+# | arg: user - the user name to create
+# | arg: pwd - the password to identify user by
+ynh_psql_create_user() {
+        ynh_psql_execute_as_root \
+        "CREATE USER ${1} WITH PASSWORD '${2}';"
```

Comment:
Postgresql est livré avec une commande buildin `createuser` qui permet justement de ne pas avoir à gérer de mot de passe et utiliser l'authentification unix. On devrait utiliser ça à la place.

Le seul cas où on peut avoir envie d'utiliser un mot de passe c'est pour php qui est trop débile pour se connecter à postgresql sans password (et là on peut avoir un helper spécial "php est trop idiot").

Reply:
While your request to use createuser is valid, to use Unix auth, we need to allow user connection, here is what our helper does (nologin):
https://github.com/YunoHost/yunohost/blob/83dca8e7c6ec4efb206140c234f51dfa5b3f3bf7/data/helpers.d/user#L55

Trying `su - user` will lead to "This account is currently not available."
So, default Postgres configuration (local all all peer => local connection for all database and all users uses unix users) is not compatible with our current user creation helper.

I understood this from: https://askubuntu.com/a/486661 (I hope I understood it well ;))

So, we'll have to use PGPASSWORD, but there is a warning:
PGPASSWORD behaves the same as the password connection parameter. Use of this environment variable is not recommended for security reasons, as some operating systems allow non-root users to see process environment variables via ps; instead consider using the ~/.pgpass file (see Section 31.15).